src/chain/optimism.rs

```diff
@@ -0,0 +1,286 @@
+            OpTxType::Eip1559 => Signed::<TxEip1559>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Eip4844 => Signed::<TxEip4844>::try_from(inner).map(OpTxEnvelope::from),
+            OpTxType::Deposit => {
+                let tx_deposit = TxDeposit {
+                    source_hash: tx
+                        .source_hash
+                        .ok_or(OptimismTransactionParsingError::MissingSourceHash)?,
+                    from: tx.inner.from,
+                    to: tx.inner.to.into(),
+                    mint: tx.mint,
+                    value: tx.inner.value,
+                    gas_limit: tx.inner.gas,
+                    is_system_transaction: tx.is_system_tx.unwrap_or_default(),
+                    input: tx.inner.input.clone(),
+                };
+                Ok(OpTxEnvelope::from(tx_deposit))
+            }
+        }
+        .map_err(|err| OptimismTransactionParsingError::ConversionError(err.to_string()))?;
+
+        let mut envelope_buf = Vec::<u8>::new();
+        tx_envelope.encode_2718(&mut envelope_buf);
+        Bytes::from(envelope_buf)
+    };
+
+    Ok(OptimismFields {
+        source_hash: tx.source_hash,
+        mint: tx.mint,
+        is_system_transaction: tx.is_system_tx,
+        enveloped_tx: Some(envelope_buf),
+    })
+}
+
+impl PevmChain for PevmOptimism {
+    type Transaction = Transaction;
+    type BlockSpecError = OptimismBlockSpecError;
+    type TransactionParsingError = OptimismTransactionParsingError;
+
+    fn id(&self) -> u64 {
+        self.id
+    }
+
+    fn build_tx_from_alloy_tx(&self, tx: alloy_rpc_types::Transaction) -> Self::Transaction {
+        Transaction {
+            inner: tx,
+            ..Transaction::default()
+        }
+    }
+
+    fn get_block_spec(&self, header: &Header) -> Result<SpecId, Self::BlockSpecError> {
+        if header.timestamp >= 1720627201 {
+            Ok(SpecId::FJORD)
+        } else if header.timestamp >= 1710374401 {
+            Ok(SpecId::ECOTONE)
+        } else if header.timestamp >= 1704992401 {
+            Ok(SpecId::CANYON)
+        } else if header.number >= 105235063 {
+            Ok(SpecId::REGOLITH)
+        } else {
+            // TODO: revm does not support when L1Block is not available
+            Err(OptimismBlockSpecError::UnsupportedSpec)
+        }
+    }
+
+    fn build_mv_memory(
+        &self,
+        hasher: &ahash::RandomState,
+        block_env: &BlockEnv,
+        txs: &[TxEnv],
+    ) -> MvMemory {
+        let beneficiary_location_hash = hasher.hash_one(MemoryLocation::Basic(block_env.coinbase));
+        let l1_fee_recipient_location_hash = hasher.hash_one(revm::L1_FEE_RECIPIENT);
+        let base_fee_recipient_location_hash = hasher.hash_one(revm::BASE_FEE_RECIPIENT);
+
+        // TODO: Estimate more locations based on sender, to, etc.
+        let mut estimated_locations = HashMap::with_hasher(BuildIdentityHasher::default());
```

Comment:
Let's add a `TODO` to benchmark without pre-estimating L1 fee & base fee recipient locations. If they are small pre-processing in sequential mode can be slower than dynamically inserting them in parallel threads.